### PR TITLE
[TOOLS-2126] Fix children worklogs distribution.

### PR DESCRIPTION
**Pull request by marcokuchla**
* Improve DSLKpi with method 'when' and Behaviors;
 * Refactoring some methods' names in DSLKpi;

### CODE REVIEW

File: kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java [hunk 1]
```diff
@@ -1,41 +1,61 @@
...
-            List<String> childrenStatuses = h.getChildrenStatus();
-            kpi.findStatus(status).ifPresent(s -> {
-                setupWorklog(kpi,s,subtasks);
-                setupWorklogs(kpi,s,childrenStatuses);
-            });
-        });
+    public void distributeWorklogs() {
+        final List<Hierarchy> hierarchies = this.hierarchies.getHierarchies();
+
+        for (Hierarchy hierarchy : hierarchies) {
```
Comment (herbert-venancio) [inline kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java]: You can use `hierarchies.getHierarchies()` directly without the local variable shadowing a field

File: kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java [hunk 2]
```diff
@@ -86,7 +86,7 @@ public Long getEffortUntilDate(String status, ZonedDateTime dateLimit) {
         return statusTransition.map(s -> s.getEffortUntilDate(dateLimit)).orElse(0l);
     }
 
-    public List<Worklog> getWorklogFromChildren(Long subtaskType) {
+    public List<Worklog> getWorklogFromChildrenTypeId(Long subtaskType) {
         return children.stream()
                 .filter(c -> c.issueType.isPresent())
```
Comment (herbert-venancio) [inline kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java]: Can merge this two filters into:
```java
                .filter(c -> c.issueType.map(type -> type.getId().equals(subtaskType)).orElse(false))
```

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java [hunk 3]
```diff
@@ -45,6 +50,35 @@ public StatusTransitionAsserter statusTransition() {
         }
         
     }
+
+    public class BehaviorFactory {
+        
+        private Map<String,IssueBehavior> issues = new LinkedHashMap<>();
+        
+        public IssueBehavior issueKpi(String pkey) {
+            issues.putIfAbsent(pkey, new IssueBehavior(DSLKpi.this.getIssueKpi(pkey)));
```
Comment (herbert-venancio) [inline kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java]: Can simplify `DSLKpi.this.getIssueKpi(pkey)` -> `getIssueKpi(pkey)`

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java [hunk 4]
```diff
@@ -45,6 +50,35 @@ public StatusTransitionAsserter statusTransition() {
...
+                this.kpi = kpi;
+            }
+            
+            public IssueBehavior appliesBehavior(DSLBehavior behavior) {
+                behavior.execute(environment, kpi);
+                return this;
+            }
+            
+            public DSLKpi then() {
+                return DSLKpi.this;
```
Comment (herbert-venancio) [inline kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java]: Or use static class:
```Java
    public static class BehaviorFactory {

        private final DSLKpi kpiContext;
        private Map<String,IssueBehavior> issues = new LinkedHashMap<>();

        public BehaviorFactory(DSLKpi kpiContext) {
            this.kpiContext = kpiContext;
        }

        public IssueBehavior issueKpi(String pkey) {
            issues.putIfAbsent(pkey, new IssueBehavior(getIssueKpi(pkey)));
            return issues.get(pkey);
        }

        public class IssueBehavior {
            private IssueKpi kpi;

            public IssueBehavior(IssueKpi kpi) {
                this.kpi = kpi;
            }

            public IssueBehavior appliesBehavior(DSLBehavior behavior) {
                behavior.execute(kpiContext.environment, kpi);
                return this;
            }

            public DSLKpi then() {
                return kpiContext;
            }
        }   
    }
```

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java (hunk 3 above)
Comment (herbert-venancio) [reply kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java]: Better yet:
```Java
            return issues.computeIfAbsent(pkey, (key) -> new IssueBehavior(kpiContext.getIssueKpi(key)));
```

File: kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java [hunk 5]
```diff
@@ -1,41 +1,61 @@
...
-        this.hierarchies.getHierarchies().stream().forEach( h -> {
-            String status = h.getFatherStatus();
-            List<Long> subtasks = h.getChildrenTypeId();
-            List<String> childrenStatuses = h.getChildrenStatus();
-            kpi.findStatus(status).ifPresent(s -> {
-                setupWorklog(kpi,s,subtasks);
-                setupWorklogs(kpi,s,childrenStatuses);
-            });
-        });
+    public void distributeWorklogs() {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java]: Fixed.

File: kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java (hunk 1 above)
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/ChildrenWorklogDistributor.java]: Fixed.

File: kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java (hunk 2 above)
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/IssueKpi.java]: Fixed.

File: kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java (hunk 3 above)
Comment (marcokuchla) [reply kpis/src/test/java/objective/taskboard/followup/kpi/enviroment/DSLKpi.java]: Fixed.

**Pull request merged**